### PR TITLE
Update storybook autocomplete component

### PR DESCRIPTION
**Pull request by aitchiss**
## What type of PR is this? (check all applicable)

- [X] Refactor

## Description

This PR reworks the MentionAutocomplete component initially created in #12831. After some more careful testing with the previous implementation we decided the Reach UI Combobox needs to fully take over the target text area. This allows us to cover a greater variety of triggers (e.g. deleting back to a previous word with `@mention`) in a more dependable way.

The `MentionAutocompleteTextArea` component now works by fully replacing the element passed to in props (copying all styles and attributes), and initialises the search behaviour when the given word (based on cursor position) begins with an `@`.

## Related Tickets & Documents

[Create mention autocomplete](https://github.com/forem/rfcs/pull/16)

## QA Instructions, Screenshots, Recordings

This can be tested in Storybook using the netlify deploy preview linked below

- Type text in the text area - you should notice no layout shifts as it is replaced with the autocomplete text area
- Experiment with adding `@mentions` in various places in the text content
- You should be able to select an option either by using down/up arrow keys and pressing enter, or by clicking an item
- After selection, the cursor should return to the end of the inserted mention
- You should also be able to just keep typing and ignore any suggestions, with the dropdown disappearing when you space away from the word
- Pressing the Escape key should hide the popup (but once you start typing again it should re-appear)
- Clicking outside of the dropdown should also close it

https://user-images.githubusercontent.com/20773163/110931920-b8f32000-8322-11eb-93c8-858066ab081d.mp4

### UI accessibility concerns?

This refactor means that the whole textarea is now assigned the role "combobox". Screen reader users are immediately informed that a list of options will become associated with the input. An `aria-live` has been added to announce to screen reader users when the `@mention` search has started populating the list of options.

It's important to note that axe flags an error for this component, as according to the [WAI ARIA guidelines](https://www.w3.org/TR/wai-aria-practices/#combobox) a combobox role should not apply to a `textarea` but a single line input. We have had a lot of discussions about how best to handle the screen reader use-case, and the implementation in this PR is very similar to the one GitHub is using for mentions. It makes sure that screen reader users can take advantage of the autocomplete functionality, despite it taking place in a textarea.

I also don't believe this implementation will degrade the current editor/comment experience for screen reader users (when added there), since in all other respects the textarea continues to operate as before. 

In short, I _think_ continuing in this direction is the best a11y experience we can offer to the best of our knowledge right now, but we should look in future to get this audited/tested by real users of assistive technology and be open to any feedback we receive.

## Added tests?

- [X] Yes (although much of the functionality will be covered by Cypress tests when implemented in the app)

## [Forem core team only] How will this change be communicated?

- [X] This change does not need to be communicated, and this is why not: This only affects the Storybook component and changes will be communicated when the component is actually implemented in the app

## [optional] What gif best describes this PR or how it makes you feel?

Baga Chipz says much betta

### CODE REVIEW

**Review by rhymes — COMMENTED** [review]
Love it! It looks easier to use than the previous version, probably because now we control the entire textarea. Does this mean, that two different components cannot coexist? Let's say we added autocompletion for emojis, would that be a sub component of this?

One last detail, why can I highlight two items? One is selected by the arrows and the other is gray because it responds to the mouse's hover:

Screenshot 2021-03-12 at 6 28 53 PM

Great work @aitchiss!

**Comment by rhymes** [comment]
@aitchiss never mind, GitHub's component works in the same way :D I thought it was a bug but apparently is how it's supposed to work

File: app/javascript/crayons/MentionAutocompleteTextArea/__stories__/mention-autocomplete.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+The `MentionAutocompleteTextArea` component uses the [Reach UI Combobox](https://reach.tech/combobox/) under the hood. It works by _replacing_ the textarea you pass in props with one enhanced with the `@mention` functionality.
+
+The autocomplete will begin fetching suggestions once a user has typed `@` plus at least two characters. A user&apos;s selection is confirmed when they either:
+
+- Click on a search option
+- Hit enter on a search option
+- Click away from the dropdown
+- Enter space to move away from the autocomplete
+
+### Mention autocomplete accessibility
```
Comment (rhymes) [inline app/javascript/crayons/MentionAutocompleteTextArea/__stories__/mention-autocomplete.md:12]: ❤️

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        return;
+      }
+
+      fetchSuggestions(searchTerm).then(({ result: fetchedUsers }) => {
+        const resultLength = Math.min(
+          fetchedUsers.length,
+          MAX_RESULTS_DISPLAYED,
+        );
+
+        const results = fetchedUsers.slice(0, resultLength);
```
Comment (rhymes) [inline app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:84]: We might want to make this smarter in a future PR, once #12975 is merged. We shouldn't ask the backend for all results and then slice them out in the frontend. We should simply ask `MAX_RESULTS_DISPLAYED` number of results right away using a limit or pagination. 

I'll make a note in #12975 to add that feature. Is there anything else you need from the backend results @aitchiss ?

cc @atsmith813

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+  const [cursorPosition, setCursorPosition] = useState(null);
+  const [ariaHelperText, setAriaHelperText] = useState('');
+
+  const isSmallScreen = useMediaQuery(`(max-width: ${BREAKPOINTS.Small}px)`);
+
+  const inputRef = useRef(null);
+
+  useEffect(() => {
+    if (searchTerm.length >= MIN_SEARCH_CHARACTERS) {
+      if (cachedSearches[searchTerm]) {
```
Comment (rhymes) [inline app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:73]: how long does this cache last? It's unlikely that it'll be an issue but I'm thinking of the scenario of a person that can't autocomplete a user's name because it didn't appear in the first page of results.

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+  const handleSelect = (username) => {
+    const textWithSelection = `${textContent.substring(
+      0,
+      selectionInsertIndex,
+    )}${username}${textContent.substring(inputRef.current.selectionStart)}`;
+
+    // Clear the current search
+    setSearchTerm('');
+    setUsers([]);
+    setAriaHelperText('');
```
Comment (rhymes) [inline app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:176]: why do we clear the aria helper text only on select and not in the other two cases where we reset the search?

**Review by rhymes — COMMENTED** [review]
Left some additional feedback from scanning the code :)

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx (hunk 2 above)
Comment (aitchiss) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:84]: Totally agree - it would be great to be able to set the max number returned in the search call. 

I don't think we need anything else from the backend results - we don't page these in or filter them at all at the moment.

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx (hunk 3 above)
Comment (aitchiss) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:73]: It will last for as long as the user is on the current page and the component's unmounted. My thinking is that the first 6 results aren't going to change from the backend for a given search term in that time (I guess it is possible a new user registers or changes their username while we're on a page, but it seems unlikely we'd be searching for them in that case).

If the user adds another character to further refine the results, then we still end up with a new search call so I feel like this is OK

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx (hunk 4 above)
Comment (aitchiss) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:176]: Good spot! I'll clear it in the other cases too 👍

**Comment by aitchiss** [comment]
>Does this mean, that two different components cannot coexist? Let's say we added autocompletion for emojis, would that be a sub component of this?

@rhymes I think we could easily expand this component in future to include another type of autocomplete. e.g. in the case of emojis, we would have another 'search indicator' - instead of `@` (probably `:`). We could then toggle behaviour based on whether the current word started with a `@` or `:` or whatever other trigger we want to listen for.

Once we start expanding that functionality it might look more like the `KeyboardShortcuts` components, e.g. passing in some options for the search behaviours like:

```jsx
{
  "@": fetchUserSuggestions,
  ":": fetchEmojiSuggestions,
  "#" fetchTagSuggestions
}
```

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx (hunk 2 above)
Comment (rhymes) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:84]: @aitchiss for now we've hardcoded the number of results in [this PR](https://github.com/forem/forem/pull/12975/files#diff-dab7116df779d917d8f1420244e812bfd0f7da4fdaaac01fb1576b925e6dbe91R4), we'll integrate a parameter if we need to and after we test the new endpoint in production.

**Review by rhymes — APPROVED** [review]
LGTM!

**Review by nickytonline — APPROVED** [review]
Tested this locally and it works great. Thanks so much for your PR @Rafi993!

File: app/assets/stylesheets/base/helpers.scss [hunk 5]
```diff
@@ -36,6 +36,15 @@
   border-top: var(--su-7) solid var(--accent-brand);
 }
 
+.screen-reader-only {
```
Comment (nickytonline) [inline app/assets/stylesheets/base/helpers.scss:39]: Nice addition. 😎

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    setTextContent(textWithSelection);
+
+    // Update the cursor to directly after the selection
+    const newCursorPosition = selectionInsertIndex + username.length + 1;
+    setCursorPosition(newCursorPosition);
+  };
+
+  useLayoutEffect(() => {
+    if (inputRef.current) {
+      const attributes = replaceElement.attributes;
```
Comment (nickytonline) [inline app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:190]: Documentation (blocking): Add a comment about what we're doing here, i.e. replacing the existing textarea.

Refactor (non-blocking): Consider putting all this logic in a function

`replaceTextArea(inputRef.current, replaceElement)`

```
function replaceTextArea(node, nodeToReplace) {
...
}
```

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+      // We need to manually remove the element, as Preact's diffing algorithm won't replace it in render
+      replaceElement.remove();
+      inputRef.current.focus();
+    }
+  }, [replaceElement]);
+
+  return (
+    <Fragment>
+      <div aria-live="polite" class="screen-reader-only">
```
Comment (nickytonline) [inline app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:211]: ♥️

**Review by nickytonline — COMMENTED** [review]
@aitchiss, I know I accidentally merged this, but a few comments. This is looking really good.
